### PR TITLE
map enums across fields

### LINKED ISSUE #990: field remappings via enum_overrides
As part of #816 we considered mapping values *between* enum fields. For example, if bond.status sometimes contains information about bond.bond_type, we should be able to easily move string values around. Here are some examples of ways to extend this functionality, in increasing order of overkill:

- `overrides.add('PENDING', BondType, BondStatus.PENDING)`
- `overrides.add('PENDING', BondType, ChargeStatus.PENDING)` (note that a charge has at most one bond, so we would know where to set things)
- `overrides.add('PENDING', BondAmount, ChargeStatus.PENDING)` (note that we don't have `BondAmount` in our code right now - the field `bond.bond_amount` has type `int` (dollars)
- `overrides.add(lambda s: s.startswith('FEE'), BondAmount, FeeDollars)` (we don't have `BondAmount` or `FeeDollars`, but imagine moving the string to a completely different field in some cases)

and to demonstrate why some of this is a bad idea:
- `overrides.add('PENDING', Gender, BondStatus.PENDING)` (not only is this super weird, we wouldn't be able to figure out which bond to set this on)

@ohinds @zacharylawrence

### PR DESCRIPTION
**Pull request by jamwalla**
close #990

### CODE REVIEW

**Comment by zacharylawrence** [comment]
@jamwalla Sorry, I didn't notice this switched from draft to ready_for_review. Should I take a look now?

**Comment by jamwalla** [comment]
No, sorry, this is broken after fixing some merge conflicts 😟. I'll let you know when it's ready!

File: recidiviz/common/constants/enum_overrides.py [hunk 1]
```diff
@@ -94,18 +94,21 @@ def build(self) -> 'EnumOverrides':
 
         def add(self,
                 label_or_predicate: Union[str, Callable[[str], bool]],
-                mapped_enum: EntityEnum) -> None:
+                mapped_enum: EntityEnum,
+                enum_class: EntityEnumMeta = None) -> None:
```
Comment (zdg2102) [inline recidiviz/common/constants/enum_overrides.py]: Maybe rename this to a named argument that would be more explicit on the call side? `from_field`?

File: recidiviz/ingest/scrape/base_scraper.py [hunk 2]
```diff
@@ -349,7 +350,15 @@ def get_enum_overrides(self) -> EnumOverrides:
         Note: Before overriding this method, consider directly adding each
         mapping directly into the respective global mappings instead.
         """
-        return EnumOverrides.empty()
+        builder = EnumOverrides.Builder()
+        for ethnicity_string in ETHNICITY_MAP:
+            # mypy is unable to correctly type the EntityEnums in
+            # constants.person. See https://github.com/python/mypy/issues/3327
+            ethnicity_enum = cast(Ethnicity, ETHNICITY_MAP[ethnicity_string])
```
Comment (zacharylawrence) [inline recidiviz/ingest/scrape/base_scraper.py]: Makes sense! Thanks for looking into this so deep!!

File: recidiviz/persistence/converter/converter_utils.py [hunk 3]
```diff
@@ -38,6 +40,25 @@ def fn(func, field_name, proto, *additional_func_args, default=None):
     return value if value is not None else default
 
 
+def get_value_from_parsed_enums(enum_type: EntityEnumMeta,
+                                parsed_enums: Set[EntityEnum],
+                                default=None):
+    matching_values = [enum for enum in parsed_enums
```
Comment (zacharylawrence) [inline recidiviz/persistence/converter/converter_utils.py]: Consider making this a `set` to allow collapsing multiple matching enums?

File: recidiviz/persistence/converter/converter_utils.py [hunk 4]
```diff
@@ -38,6 +40,25 @@ def fn(func, field_name, proto, *additional_func_args, default=None):
...
+                       if isinstance(enum, enum_type)]
+    if len(matching_values) > 1:
+        raise ValueError("Found multiple values for enum field %s: %s" %
+                         (enum_type, matching_values))
+    if len(matching_values) == 1:
+        return matching_values[0]
+    return default
+
+
+def parse_enums(proto, enum_fields: Dict[str, Callable],
```
Comment (zacharylawrence) [inline recidiviz/persistence/converter/converter_utils.py]: Can `proto` have a type too?

File: recidiviz/persistence/converter/charge.py [hunk 5]
```diff
@@ -33,28 +33,37 @@ def copy_fields_to_builder(
...
+    enum_fields = {
+        'degree': ChargeDegree.parse,
+        'charge_class': ChargeClass.parse,
+        'status': ChargeStatus.parse,
+        'court_type': CourtType.parse
+    }
+    parsed_enums = parse_enums(proto, enum_fields,
+                               metadata.enum_overrides)
+
+    new.degree = get_value_from_parsed_enums(ChargeDegree, parsed_enums)
```
Comment (zacharylawrence) [inline recidiviz/persistence/converter/charge.py]: Consider making this a class so you can say something like: `parsed_enums.get(ChargeDegree)`?

File: recidiviz/ingest/scrape/base_scraper.py [hunk 6]
```diff
@@ -349,7 +350,15 @@ def get_enum_overrides(self) -> EnumOverrides:
...
+            if ethnicity_enum != Ethnicity.EXTERNAL_UNKNOWN:
```
[the remark this answers is not in the document]
Comment (jamwalla) [reply recidiviz/ingest/scrape/base_scraper.py]: cool!

File: recidiviz/persistence/converter/converter_utils.py [hunk 7]
```diff
@@ -38,6 +40,25 @@ def fn(func, field_name, proto, *additional_func_args, default=None):
     return value if value is not None else default
 
 
+def get_value_from_parsed_enums(enum_type: EntityEnumMeta,
+                                parsed_enums: Set[EntityEnum],
+                                default=None):
+    matching_values = [enum for enum in parsed_enums
+                       if isinstance(enum, enum_type)]
+    if len(matching_values) > 1:
```
[the remark this answers is not in the document]
Comment (jamwalla) [reply recidiviz/persistence/converter/converter_utils.py]: done

File: recidiviz/persistence/converter/converter_utils.py [hunk 8]
```diff
@@ -38,6 +40,25 @@ def fn(func, field_name, proto, *additional_func_args, default=None):
     return value if value is not None else default
 
 
+def get_value_from_parsed_enums(enum_type: EntityEnumMeta,
```
[the remark this answers is not in the document]
Comment (jamwalla) [reply recidiviz/persistence/converter/converter_utils.py]: yes

File: recidiviz/persistence/converter/converter_utils.py (hunk 4 above)
Comment (jamwalla) [reply recidiviz/persistence/converter/converter_utils.py]: this is `Union[ingest_info_pb2.Person, ingest_info_pb2.Booking]`... etc so we decided against

**Pull request merged**